### PR TITLE
IGNITE-8982: Reuse H2 connections

### PR DESCRIPTION
**Pull request by pavlukhin**
Simple object pool introduced and instance of pool is created per thread. Connection and cached statements for it are stored in compound object in pool.

### CODE REVIEW

**Comment by pavlukhin** [comment]
I added some tests for `ObjectPool`. But it seems they do not conform general approach. I will try to figure how to structure tests properly.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java [hunk 1]
```diff
@@ -18,25 +18,31 @@
...
 
+    /** */
+    private volatile H2StatementCache statementCache;
+
     /**
      * @param conn Connection to use.
      */
     H2ConnectionWrapper(Connection conn) {
         this.conn = conn;
+        clearStatementCache();
```
Comment (gvvinblade) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java]: looks weird to clear cache just after the object creation; isn't it more descriptive to just assign new value to the field?

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java [hunk 2]
```diff
@@ -60,8 +66,28 @@ public Connection connection() {
...
+     */
+    public H2StatementCache statementCache() {
+        return statementCache;
+    }
+
+    /**
+     * Clears statement cache.
+     */
+    public void clearStatementCache() {
+        statementCache = new H2StatementCache(256);
```
Comment (gvvinblade) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java]: Try not to use any "magic numbers"

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ObjectPool.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
+package org.apache.ignite.internal.processors.query.h2;
```
Comment (gvvinblade) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ObjectPool.java]: Apache header

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/UpdatePlan.java [hunk 4]
```diff
@@ -630,20 +632,17 @@ private AbstractIterator(IgniteH2Indexing idx, QueryCursor<List<?>> cur, UpdateP
...
         /** {@inheritDoc} */
-        @Override protected void onClose() throws IgniteCheckedException {
+        @Override protected void onClose() {
             cur.close();
 
-            Connection conn0 = conn;
-
-            if (conn0 != null)
-                U.close(conn0, null);
+            conn.recycle();
```
Comment (gvvinblade) [inline modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/UpdatePlan.java]: Possible NPE

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java (hunk 1 above)
Comment (pavlukhin) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java]: Motivation was in code reuse. Do you like it more with private `initStatementCache` method which is called from constructor and `clearStatementCacheMethod`?

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java (hunk 2 above)
Comment (pavlukhin) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java]: Actually, it is cache size. And it is used only in one place. I do not see much profit if it is extracted as constant. But it does not make much sense. I can keep it as is or extract a constant.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ObjectPool.java (hunk 3 above)
Comment (pavlukhin) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ObjectPool.java]: Yep. Thanks! I should configure copyright settings.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java (hunk 1 above)
Comment (pavlukhin) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java]: private `initStatementCache` method introduced

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java (hunk 2 above)
Comment (pavlukhin) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ConnectionWrapper.java]: Constant extracted.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/UpdatePlan.java (hunk 4 above)
Comment (pavlukhin) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/UpdatePlan.java]: Thanks! Fixed.

File: modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ObjectPool.java (hunk 3 above)
Comment (pavlukhin) [reply modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/ObjectPool.java]: Header added.

**Comment by pavlukhin** [comment]
Also I reformatted tests in JUnit 3 style and added them to `IgniteCacheQuerySelfTestSuite`.

**Pull request closed without merging**